core/src/main/scala/cilib/Boundary.scala

```diff
@@ -0,0 +1,82 @@
+package cilib
+
+import scalaz._
+import Scalaz._
+import spire.math.Interval
+
+object Boundary {
+
+  final case class Enforce[F[_], A](f: (A, Interval[Double]) => F[A])
+  final case class EnforceTo[F[_], A](f: (A, Interval[Double], A) => F[A])
+
+  def enforce[F[_]: Applicative, A: spire.math.Numeric](x: Position[A],
+                                                        f: Enforce[F, A]): F[NonEmptyList[A]] =
+    x.pos.zip(x.boundary).traverse { case (a, b) => f.f(a, b) }
+
+  def enforceTo[F[_], A: spire.math.Numeric, B](x: Position[A], z: Position[A], f: EnforceTo[F, A])(
+      implicit F: Applicative[F]): F[NonEmptyList[A]] =
+    x.pos.zip(x.boundary).zip(z.pos).traverse {
+      case ((a, b), c) => f.f(a, b, c)
+    }
+
+  def clamp[A](implicit N: spire.math.Numeric[A]) =
+    absorb
+
+  def absorb[A](implicit N: spire.math.Numeric[A]) =
+    Enforce((a: A, b: Interval[Double]) =>
+      Need {
+        val z = N.toDouble(a)
+
+        if (z < b.lowerValue) N.fromDouble(b.lowerValue)
+        else if (z > b.upperValue) N.fromDouble(b.upperValue)
+        else a
+    })
+
+  def random[A](implicit N: spire.math.Numeric[A]) =
+    Enforce(
+      (a: A, b: Interval[Double]) =>
+        if (b.contains(N.toDouble(a))) RVar.pure(a)
+        else Dist.uniform(b).map(N.fromDouble))
+
+  def reflect[A](implicit N: spire.math.Numeric[A]) =
+    Enforce((a: A, b: Interval[Double]) =>
+      Need {
+        val z = N.toDouble(a)
+
+        if (z < b.lowerValue) N.fromDouble(b.lowerValue - (z - b.lowerValue))
+        else if (z > b.upperValue) N.fromDouble(b.upperValue - (z - b.upperValue))
+        else a
+    })
+
+  def wrap[A](implicit N: spire.math.Numeric[A]) =
+    toroidal
+
+  def toroidal[A](implicit N: spire.math.Numeric[A]) =
+    Enforce((a: A, b: Interval[Double]) =>
+      Need {
+        val z = N.toDouble(a)
+
+        if (z < b.lowerValue) N.fromDouble(b.lowerValue + (z - b.lowerValue))
+        else if (z > b.upperValue) N.fromDouble(b.upperValue + (z - b.upperValue))
```

Comment:
If I understand this correctly, shouldn't the logic be 
```if (z < lowerValue) upperValue - (z - lowerValue)
if (z > upperValue) lowerValue  - (z - upperValue)
else z```
You have to come from the other side instead of the same side it has gone over.

Reply:
Yip, exactly right - it does an environment "wrap", kinda like the old platform games.